libs/constrained-generators/src/Constrained/Univ.hs

```diff
@@ -0,0 +1,313 @@
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE FunctionalDependencies #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImportQualifiedPost #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Constrained.Univ where
+
+import Constrained.List
+import Control.Monad.Identity
+import Data.Kind
+import Data.Functor.Const
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set (Set)
+import Data.Set qualified as Set
+
+class FunctionLike fn where
+  -- | The semantics of a function is given by `sem`
+  sem :: fn as b -> FunTy as b
+
+class Member (fn :: [Type] -> Type -> Type) univ where
+  -- | Inject a specific function symbol into a function
+  -- universe.
+  injectFn :: fn as b -> univ as b
+
+------------------------------------------------------------------------
+-- Fixpoints
+------------------------------------------------------------------------
+
+{-
+On the most basic level what we are after is a first order representation of
+typed functions. Furthermore, we need the representation to be compositional so
+that you can plug-in your own functions into the universe.
+
+That might sound like it's no problem, but the issue is with higher-order
+functions. Consider:
+
+data ListFn where
+  MapFn :: _ '[a] b -> ListFn '[[a]] [b]
+
+What should go in place of `_`? It turns out that you need to specify the global
+function universe and make sure the global universe is the same everywhere.
+
+data ListFn univ where
+  MapFn :: univ '[a] b -> ListFn univ '[[a]] [b]
+
+-}
+
+newtype Fix fn (as :: [Type]) b = Fix (fn (Fix fn) as b)
+
+instance Member fn (fn' (Fix fn')) => Member fn (Fix fn') where
+  injectFn = Fix . injectFn
+
+deriving via (fn (Fix fn)) instance FunctionLike (fn (Fix fn)) => FunctionLike (Fix fn)
```

Comment:
Just to test my intuition here, would `deriving newtype` here give the same result?

Reply:
Possibly. But I don't know precisely how `deriving newtype` works.